cscs-checks/prgenv/gpu_direct_acc.py

```diff
@@ -1,46 +1,52 @@
 import reframe.utility.sanity as sn
-from reframe.core.pipeline import RegressionTest
-
+from reframe.core.launchers import LauncherWrapper
 
-class GpuDirectAccCheck(RegressionTest):
+@rfm.simple_test
+class GpuDirectAccCheck(rfm.RegressionTest):
     def __init__(self, **kwargs):
-        super().__init__('gpu_direct_acc_check',
-                         os.path.dirname(__file__), **kwargs)
+        super().__init__()
+        self.descr = 'tests gpu-direct for Fortran OpenACC'
         self.valid_systems = ['daint:gpu', 'dom:gpu', 'kesch:cn']
-        self.valid_prog_environs = ['PrgEnv-cray']
+        self.valid_prog_environs = ['PrgEnv-cray*', 'PrgEnv-pgi*']
         if self.current_system.name in ['daint', 'dom']:
             self.modules = ['craype-accel-nvidia60']
-            self._pgi_flags = '-acc -ta=tesla:cc60'
+            self._pgi_flags = '-acc -ta=tesla:cc60 -Mnorpath'
             self.variables = {'MPICH_RDMA_ENABLED_CUDA': '1'}
             self.num_tasks = 2
             self.num_gpus_per_node = 1
             self.num_tasks_per_node = 1
         elif self.current_system.name in ['kesch']:
             self.modules = ['craype-accel-nvidia35']
-            self._pgi_flags = '-acc -ta=tesla:cc35'
-            self.variables = {'MPICH_RDMA_ENABLED_CUDA': '1',
-                              'MV2_USE_CUDA': '1',
-                              'MV2_USE_GPUDIRECT': '1',
-                              'G2G': '1'}
+            self._pgi_flags = '-acc -ta=tesla:cc35 -Mnorpath'
+            self.variables = {
+                'MPICH_RDMA_ENABLED_CUDA': '1',
+                'MV2_USE_CUDA': '1',
+                'G2G': '1',
+                'X': '$(pkg-config --variable=libdir mvapich2-gdr)'
+            }
             self.num_tasks = 8
             self.num_gpus_per_node = 8
             self.num_tasks_per_node = 8
 
         self.sourcepath = 'gpu_direct_acc.F90'
-        self.sanity_patterns = sn.assert_found(r'Result :\s+OK', self.stdout)
+        self.sanity_patterns = sn.all([
+            sn.assert_found(r'GPU with OpenACC', self.stdout),
+            sn.assert_found(r'Result :\s+OK', self.stdout)
+        ])
+        self.launch_options = []
         self.maintainers = ['AJ', 'VK']
         self.tags = {'production'}
 
     def setup(self, partition, environ, **job_opts):
-        if environ.name == 'PrgEnv-cray':
+        if environ.name.startswith('PrgEnv-cray'):
             environ.fflags = '-hacc -hnoomp'
-        elif environ.name == 'PrgEnv-pgi':
+        elif environ.name.startswith('PrgEnv-pgi'):
             environ.fflags = self._pgi_flags
 
         super().setup(partition, environ, **job_opts)
-
-
-def _get_checks(**kwargs):
-    return [GpuDirectAccCheck(**kwargs)]
+        if (self.current_system.name in ['kesch']) and \
+            (environ.name.startswith('PrgEnv-pgi')):
+            self.job.launcher = LauncherWrapper(self.job.launcher,
+                'LD_PRELOAD=$X/libmpi.so', self.launch_options
```

Comment:
@ajocksch Why an `export LD_PRELOAD=...` wouldn't work? You could simply set `LD_PRELOAD` in `self.variables`.